### DISCUSSION TITLE
How to play/pause particles animation?

### DISCUSSION THREAD

**Discussion by gbdistro** [tsparticles/react#108] [Q&A]
Hello, I'm wondering if there's a way to control the particles animation state in a react environment. I'm using `tsparticles v3.3.0` and `@tsparticles/react v3.0.0`. I have two buttons that each have an `onClick` handler that is responsible or either playing or pausing the particles animation.

I want to achieve the same play/pause button functionality that the [particles.js.org](https://particles.js.org/) config editor achieves. My buttons work sometimes, but the results are very inconsistent. More often than not, the handlers are either ignored or produce weird side-effects.

I've tried looking for solutions all over the `tsparticles` umbrella. The closest solution I found was [Issue #4697](https://github.com/tsparticles/tsparticles/issues/4697). The problem with this (and other related) solutions is that they are either out of date or aren't using an official framework component (such as [Issue #5231](https://github.com/tsparticles/tsparticles/issues/5231)). The `v3.0.0` components don't have access to the same props, such as the `particlesInit` prop, though I'm not sure how much this really matters.

I have created a [CodeSandbox example](https://codesandbox.io/p/devbox/tsparticles-react-testing-mnpk27?file=%2Fcomponents%2FHexagonParticles.tsx) that goes into more detail. See the README and the content on the rendered home page. This isn't really a minimal sandbox, but it emulates my exact environment. The only file that matters is the `<HexagonParticles />` component found in `'./components/HexagonParticles.tsx'`. It is setup nearly exactly as the [example react app](https://github.com/tsparticles/react/blob/main/apps/react/src/App.js) located in this package.

I'm not sure if other frameworks have a hard time with this functionality. It could be a react thing due to how re-rendering works, and that I simply lack the knowledge on how to overcome this.

Any help or insight is much appreciated.

**Comment by matteobruni** [comment]
I've checked the CodeSandbox, and I noticed an issue immediately:

```
    initParticlesEngine(async (engine) => {
      await loadFull(engine);
      await loadPolygonPath(engine);
    }).then(() => {
      setInit(true);
    });
```

This is better if it's moved in the App.tsx file or something like that, these plugins must be loaded only once per application. The rest seems fine, maybe this issue could have unexpected behavior since every plugin could reload the instance.

**Comment by gbdistro** [comment]
Yep, making sure to only load the engine once is the key.
It turns out that `React.StrictMode` was causing the issue. It makes the `useEffect` that initializes the engine to run twice, which must've been causing the problem. I disabled `React.StrictMode` and now the button handlers work as expected. This means it's just something to look out for during development, since `React.StrictMode` has no effect in production.

Thank you @matteobruni for your insight and for this really cool library. I'm looking forward to applying it to more use cases!

---

For anyone else that found their way here, this is how I handled it. You can also visit the CodeSandbox link I posted above.
In `next.config.js`:
```js
const nextConfig = {
  reactStrictMode: false,
  // other configs...
};
```
I created a `<ParticlesProvider>` component that wraps everything in the `<body>` in the root `layout.tsx`. This provider holds the `useEffect` that loads the engine. I created a React Context that is subscribed to the `isInitialized` status of the provider, and made a `useParticles` hook that returns the value of the context. Any (client) component can use this hook to conditionally render the `<Particles />` based on the `isInitialized` value.

This approach allows the engine to load in the background. If there's a `<Particles />` component on another page then there's a chance it will already be playing by the time the user reaches it. You can also have many `<Particles />` throughout the app if you so desire.

The new challenge I have is figuring out how to display a play *or* pause button depending on an `isPlaying` state. Any attempt to set state within the button's `onClick` handler causes the component to re-render and restarts the `<Particles />` animation. I believe this is a problem I'll need to solve, but if anyone happens to know the answer then I'll gladly accept the help.

**Reply by srsholmes** [reply]
Did you figure out the re-rendering issue? 

I've been trying to use this library for days now and it's not really useable within a react app if you cannot re-render the / re-use the container and call the container functions programatically. 

I've tried a tonne of different ways, and havn't figured out a way to reliably re-use a container if the components get re-rendered.

**Reply by gbdistro** [reply]
@srsholmes

I did end up solving my particular re-rendering issue shortly after posting my answer/reply, and the code is there in the [codesandbox link](https://codesandbox.io/p/devbox/tsparticles-react-testing-mnpk27?file=%2Fcomponents%2FTunnelParticles.tsx%3A8%2C45) within the `TunnelParticles.tsx` component.

The way I solved it was by wrapping the `<Particles />` in a `useMemo` inside of the `<TunnelParticles />` component:

```tsx
// This component has been shortened for this comment. Check the codesandbox link for the full file.
export default function TunnelParticles() {
  // states and stuff...

  const memoizedParticles = useMemo(() => {
    return (
      <Particles
        id="tsparticles-fireworks" // ← non-critical typo from copy-pasting, "tsparticles-tunnel" would've been more fitting.
        particlesLoaded={particlesLoaded}
        options={tunnelParticles}
        className="col-start-1 row-start-1 max-h-[calc(100vh-70px)]"
      />
    );
    // eslint-disable-next-line
  }, []);

  // handlers stuff...

  if (isInitialized) {
    return (
      <div>
        {memoizedParticles}
      </div>
    )
  }

  return ( /* ... */ );
}
```

Unfortunately this isn't fresh on my mind anymore, so I'm not sure of the other methods I tried using to make this work. Memoization issues are sometimes solved without `useMemo` by separating the 'data-to-memoize' into its own file. This is the approach I use with Tanstack Table due to encountering the issues the Tanstack team describe [here](https://tanstack.com/table/latest/docs/faq). There might have been a reason why I didn't take that approach with `ts-particles`, but again, I don't really remember anymore.

React 19 is just around the corner, and they say that the [new compiler](https://react.dev/learn/react-compiler) will make the manual usage of `memo`, `useMemo`, and other hooks/APIs unnecessary. I'm curious to see how that works out.
